### ISSUE TITLE
about the shape and mask

### ISSUE THREAD

**Issue by DCdream** [HKUST-KnowComp/R-Net#36]
with tf.variable_scope("match"):
            self_att = dot_attention(
                att, att, mask=self.c_mask, hidden=d, keep_prob=config.keep_prob, is_train=self.is_train)
            rnn = gru(num_layers=1, num_units=d, batch_size=N, input_size=self_att.get_shape(
            ).as_list()[-1], keep_prob=config.keep_prob, is_train=self.is_train)
            match = rnn(self_att, seq_len=self.c_len)
in this variable_scope the  shape of "att" is [batch_size, c_max_len, d]
the shape of mask is [batch_size, c_max_len]. 
Is that right

**Comment by burglarhobbit** [comment]
yes, that is right. because the last index of the shape of 'att' is the feature size 'd' of each component from 0 to (c_max_len - 1). so mask will be broadcasted on the last dimension like this:

Internally by TensorFlow, `[[True, False, True],[False, True, True]]` will be broadcasted to ->  `[[[True]*d, [False]*d, [True]*d],[[False]*d, [True]*d, [True]*d]]` where *d means that particular list will be repeated 'd' times.

**Comment by DCdream** [comment]
with tf.variable_scope("attention"):
            inputs_ = tf.nn.relu(
                dense(d_inputs, hidden, use_bias=False, scope="inputs"))
            #inputsW
            memory_ = tf.nn.relu(
                dense(d_memory, hidden, use_bias=False, scope="memory"))
            # inputsW
            outputs = tf.matmul(inputs_, tf.transpose(
                memory_, [0, 2, 1])) / (hidden ** 0.5)
            #相当于求每个时刻各个向量的score[b, t, t]
            mask = tf.tile(tf.expand_dims(mask, axis=1), [1, JX, 1])
OK!
so in the function dot_attention(), and in the .variable_scope("attention"), the outputs became the shape of [batch_size, c_maxlen, c_max_len], which means the unnormalized atention score, and the mask also became the shape of [batch_size, c_maxlen, c_maxlen]. Is that all right.
And i don't understand where is the operation like this "Internally by TensorFlow, [[True, False, True],[False, True, True]] will be broadcasted to -> [[[True]*d, [False]*d, [True]*d],[[False]*d, [True]*d, [True]*d]] where *d means that particular list will be repeated 'd' times." in the source code.

**Comment by burglarhobbit** [comment]
https://www.tensorflow.org/api_docs/python/tf/multiply

In the link, they have mentioned: 
```
NOTE: `tf.multiply` supports broadcasting.
```

**Comment by DCdream** [comment]
why closed？

________________________________
发件人: Wenxuan Zhou <notifications@github.com>
发送时间: 2018年4月23日 9:18:32
收件人: HKUST-KnowComp/R-Net
抄送: DCdream; Author
主题: Re: [HKUST-KnowComp/R-Net] about the shape and mask (#36)


Closed #36<https://github.com/HKUST-KnowComp/R-Net/issues/36>.

―

**Comment by wzhouad** [comment]
Haven't your problem been solved?
